Fix/devdocs component name

sirreal opened the pull request:
#### Changes proposed in this Pull Request

In #38839 I noticed many devdocs components don't display a proper component name (when production optimizations are enabled).

- Ensure human readable names are used in `/devdocs/design`. This greatly improves search.
- Fixes some issues with some components causing issues in our devdocs setup. Gridicons had this issue, I added wrapper components as needed.
- Sorts the components alphabetically.

##### Not helpful 😕 

![before](https://user-images.githubusercontent.com/841763/72802169-7caabf00-3c4b-11ea-876d-c17c225401e3.png)

##### Much better 😁✨ 


#### Testing instructions

* Visit [`/devdocs/design`](https://calypso.live/devdocs/design?branch=fix/devdocs-component-name) and compare with [wpcalypso](https://wpcalypso.wordpress.com/devdocs/design). You should see several fixed component names.

sirreal committed:
Add displayName

sirreal committed:
Fix displayNames for devdocs

sirreal committed:
Sort components

sirreal commented:
An additional odd one not fixed here: [`<Feature />`](http://wpcalypso.wordpress.com/devdocs/design/feature) looks fine in development, but not in production.

## Prod

![Screen Shot 2020-01-21 at 12 45 46](https://user-images.githubusercontent.com/841763/72802387-f80c7080-3c4b-11ea-97ab-338d067ee315.png)

## Dev

![Screen Shot 2020-01-21 at 12 56 11](https://user-images.githubusercontent.com/841763/72803244-e5933680-3c4d-11ea-9dc6-120dfa185b7e.png)

sirreal commented:
> An additional odd one not fixed here: [`<Feature />`](http://wpcalypso.wordpress.com/devdocs/design/feature) looks fine in development, but not in production.

Handling in #38971

jsnajdr commented on client/components/accordion/docs/example.jsx:
```diff
@@ -10,15 +10,19 @@ import React from 'react';
+const Accordion = props => <AccordionComponent { ...props } />;
 Accordion.displayName = 'Accordion';
+const Gridicon = props => <GridiconComponent { ...props } />;
 Gridicon.displayName = 'Gridicon';
```
Do we really need to give a `displayName` to the `Gridicon` or `Accordion` component itself? To make devdocs UI show correct names, I think only the `AccordionExample` component should have a `displayName`?

sirreal replied:
That did seem to be required for this component to work correctly. It's necessary for `GridIcon` to render here:

![Screen Shot 2020-01-21 at 13 34 06](https://user-images.githubusercontent.com/841763/72805299-d662b780-3c52-11ea-9aed-17b487f08d53.png)

and I recall it throwing errors as I worked on these fixes.

sirreal replied:
I think that's related to the way the playground works.

jsnajdr replied:
This is strange. `react-live` needs to somehow map a component name you type into the editor (i.e., `<JetpackLogo />` or `<Button />`) onto the minified name. That should be done by the `playground-scope` module and the `<LiveProvider scope>` prop.

If that didn't work reliably, it would mean that all components that are supported in the playground (see the list of reexports in `playground-scope`) would need a `displayName` field.

Could you try to reproduce what exactly were the errors thrown when these components wrappers weren't present?

jsnajdr replied:
Update: I think I see these errors (failing to render `Gridicon`) now in my local Calypso.

sirreal replied:
I'd speculate the names come from here:

https://github.com/Automattic/wp-calypso/blob/c528da6820b6c81a7aecfc6dbf3a0dc780d01349/client/devdocs/design/playground-scope.js#L75

Perhaps we should be using those playground scoped components in our examples?

jsnajdr replied:
The trouble with `Gridicon` is that it's a return value of `React.memo( React.forwardRef( ... ) )` and then the `react-element-to-jsx-string` library can't figure out the display name of such a thing.

The library's `format` function accepts a custom `displayName` function as an option. And we could solve the problem with a custom implementation that takes `playground-scope` into account. I'll try to do it.

jeraldjuice commented on client/components/gsuite/docs/new-user-list.tsx:
```diff
@@ -92,4 +92,6 @@ const GSuiteNewUserListExample = () => {
 	);
 };
 
+GSuiteNewUserListExample.displayName = 'GSuiteNewUserListExample';
```
This should go on the wrapper component (`./gsuite/docs/example`) instead. :)

<img width="534" alt="Screen Shot 2020-01-22 at 9 11 22 PM" src="https://user-images.githubusercontent.com/8892849/72951522-a6074000-3d5c-11ea-865c-0a7366378951.png">

sirreal replied:
Indeed, thanks! Pushed a fix.

jeraldjuice reviewed (dismissed):
This looks good - the Gsuite component just needs attention.

sirreal committed:
Fix gsuite name

sirreal commented:
Rebased and fixed GSuite example.

jsnajdr committed:
Devdocs Playground: use playground-scope to infer displayName in example code

Reliably get the display name of a JSX element even in minified bundle and also for
elements that are results of `React.memo` or `React.forwardRef`. These are not classical
functions with names (functional components or class constructors) and the display name
is not normally possible to figure out.

jsnajdr committed:
Don't create named wrappers for components in devdocs examples

jsnajdr committed:
FeatureExample: refactor the component and the example's displayName

jsnajdr commented:
@sirreal I pushed three new commits. The function that formats JSX markup as string now has custom callback to figure out the display name. That removes need for creating custom named wrappers for components like `Button` or `Gridicon`.

Then the only `displayName` updates we need in this PR are the `*Example` components that are indexed by devdocs.

Let me know if everything still works for you as expected.

sirreal commented on client/components/accordion/docs/example.jsx:
```diff
@@ -13,12 +13,10 @@ import React from 'react';
+export default function AccordionExample( props ) {
 	return props.exampleCode;
 }
+AccordionExample.displayName = 'AccordionExample';
```
@jsnajdr This is now redundant because the function name can be inferred?

sirreal replied:
Nevermind, I misinterpreted some of your changes. I think we'll still need these for wpcalypso production env with mangled names.

sirreal commented on client/components/accordion/docs/example.jsx:
```diff
@@ -13,12 +13,10 @@ import React from 'react';
+export default function AccordionExample( props ) {
 	return props.exampleCode;
 }
+AccordionExample.displayName = 'AccordionExample';
```
```suggestion
```

sirreal commented:
> @sirreal I pushed three new commits. The function that formats JSX markup as string now has custom callback to figure out the display name. That removes need for creating custom named wrappers for components like `Button` or `Gridicon`.

Much nicer 👍 This is working as expected, thanks!

jsnajdr reviewed (approved):
This seems to work great, including the playground code samples :+1:
